presto-delta/src/main/java/com/facebook/presto/delta/DeltaPageSourceProvider.java

```diff
@@ -0,0 +1,332 @@
+import static com.google.common.base.Strings.nullToEmpty;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static org.apache.parquet.io.ColumnIOConverter.constructField;
+
+public class DeltaPageSourceProvider
+        implements ConnectorPageSourceProvider
+{
+    private final HdfsEnvironment hdfsEnvironment;
+    private final TypeManager typeManager;
+    private final FileFormatDataSourceStats fileFormatDataSourceStats;
+
+    @Inject
+    public DeltaPageSourceProvider(
+            HdfsEnvironment hdfsEnvironment,
+            TypeManager typeManager,
+            FileFormatDataSourceStats fileFormatDataSourceStats)
+    {
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.fileFormatDataSourceStats = requireNonNull(fileFormatDataSourceStats, "fileFormatDataSourceStats is null");
+    }
+
+    @Override
+    public ConnectorPageSource createPageSource(
+            ConnectorTransactionHandle transactionHandle,
+            ConnectorSession session,
+            ConnectorSplit split,
+            ConnectorTableLayoutHandle layout,
+            List<ColumnHandle> columns,
+            SplitContext splitContext)
+    {
+        DeltaSplit deltaSplit = (DeltaSplit) split;
+        DeltaTableLayoutHandle deltaTableLayoutHandle = (DeltaTableLayoutHandle) layout;
+        DeltaTableHandle deltaTableHandle = deltaTableLayoutHandle.getTable();
+
+        HdfsContext hdfsContext = new HdfsContext(
+                session,
+                deltaSplit.getSchema(),
+                deltaSplit.getTable(),
+                deltaSplit.getFilePath(),
+                false);
+        Path filePath = new Path(deltaSplit.getFilePath());
+        List<DeltaColumnHandle> deltaColumnHandles = columns.stream()
+                .map(DeltaColumnHandle.class::cast)
+                .collect(Collectors.toList());
+
+        return createParquetPageSource(
+                hdfsEnvironment,
+                session.getUser(),
+                hdfsEnvironment.getConfiguration(hdfsContext, filePath),
+                filePath,
+                deltaSplit.getStart(),
+                deltaSplit.getLength(),
+                deltaSplit.getFileSize(),
+                deltaColumnHandles,
+                deltaTableHandle.toSchemaTableName(),
+                isFailOnCorruptedParquetStatistics(session),
+                getParquetMaxReadBlockSize(session),
+                isParquetBatchReadsEnabled(session),
+                isParquetBatchReaderVerificationEnabled(session),
+                typeManager,
+                deltaTableHandle.getPredicate(),
+                fileFormatDataSourceStats);
+    }
+
+    private static ConnectorPageSource createParquetPageSource(
+            HdfsEnvironment hdfsEnvironment,
+            String user,
```

Comment:
one following TODO: we should try put Parquet page source code in utilities, so that could reuse them without copying code

Reply:
Yep, I see this code is copied in three places. Planning to have a follow up PR to refactor this. There are few places where connector dependent code is there, but those can be abstracted.